design-proposals/klip-4-produce-data.md

```diff
@@ -0,0 +1,201 @@
+complex types. Since we know the schema of the inserts, we can initially load the values into a 
+connect data type. We already have the proper serializers to convert from connect to each of the
+supported serialization formats.
+
+## Test plan
+
+This functionality is expected to work both in Interactive as well as Headless modes, and therefore
+all tests described below should be done in both:
+
+* Test that values can be serialized into JSON/AVRO/DELIMITED
+* Test that schemas are properly verified
+* Test `NULL` and table tombstones
+* Test `CREATE STREAM/TABLE` with topics that don't exist
+* Test producing data in topics that already exist
+* Test producing data to topics that are not associated with a KSQL source
+
+## Documentation Updates
+
+* The `CREATE TABLE` and `CREATE STREAM` syntax references will have an updated table properties
+section that includes the following rows:
+> ```rst
+> +=========================+======================================================================================================+
+> | KAFKA_TOPIC (required)  | The name of the Kafka topic that backs this source. The topic must either already exist in Kafka, or |
+> |                         | PARTITIONS and REPLICAS must be specified to create the topic. Command will fail if the topic exists |
+> |                         | with different partition/replica counts.                                                             |
+> +-------------------------+------------------------------------------------------------------------------------------------------+
+> | PARTITIONS              | The number of partitions in the backing topic. This property must be set if creating a SOURCE without|
+> |                         | an existing topic.                                                                                   |
+> +-------------------------+------------------------------------------------------------------------------------------------------+
+> | REPLICAS                | The number of replicas in the backing topic. This property must be set if creating a SOURCE without  |
+> |                         | an existing topic.                                                                                   |
+> +-------------------------+------------------------------------------------------------------------------------------------------+
+> ```
+* A new section for `INSERT INTO ... VALUES` will be added to the `sytnax.rst`:
+> ```rst
+> **Synopsis**
+> 
+> .. code:: sql
+>
+>   INSERT INTO stream_name/table_name [( column_name [, ...] )]
+>     VALUES (value [, ...])
+> 
+> **Description**
+> 
+> Produce a row into an existing stream or table and its underlying topic based on
+> explicitly sepcified values.
+>
+> Column values can be given in several ways:
+> * Any column not explicitly given a value is set to its default (explicit or implicit) value.
```

Comment:
What are "explicit" and "implicit" default values? Are explicit defaults only for Avro? What are implicit defaults? 0 for numerics and the empty string for strings? Or is the implicit default always null, since all KSQL schema fields are optional anyway?

I'm not sold on the value of allowing users to only specify a subset of columns to populate when producing data. Seems to complicate the design a fair bit, for a tool that's meant for development/testing which typically aren't done with massive schemas anyway.

Reply:
Good catch here, definitely needs flushing out a little. I think we should follow SQL standard here, which is to allow partial schemas, but to make it simple I think the default for everything should, indeed, be `null`. 

In the future, we could support setting an explicit default value for a column using standard SQL:
```sql
CREATE TABLE Persons (
    ID int NOT NULL,
    LastName varchar(255) NOT NULL,
    FirstName varchar(255),
    Age int,
    City varchar(255) DEFAULT 'Sandnes'
);
```
source: https://www.w3schools.com/sql/sql_default.asp

Also if we support primitive values (might be useful) we would need to come up with sane defaults for those if not supplied. Thoughts?

Reply:
> I think we should follow SQL standard here, which is to allow partial schemas, but to make it simple I think the default for everything should, indeed, be null.

I didn't realize the SQL standard allows inserting using partial schemas -- if that's the case then I'm definitely sold on supporting it in KSQL as well.

I agree defaulting to `null` is easier behavior to understand, but maybe others used to using explicit defaults in Avro would disagree.